pioinstaller/penv.py

```
@@ -61,15 +92,86 @@
+        with open(file_path, "rb") as f:
+            for chunk in iter(lambda: f.read(4096), b""):
+                sha256_hash.update(chunk)
+        calculated_hash = sha256_hash.hexdigest()
+        is_valid = calculated_hash == expected_sha256
+        if not is_valid:
+            log.error("Checksum mismatch for %s: expected %s, got %s",
+                      os.path.basename(file_path), expected_sha256,
+                      calculated_hash)
+        else:
+            log.debug("Checksum verified for %s", os.path.basename(file_path))
+        return is_valid
```

Ruff (0.12.2)

Consider moving this statement to an `else` block

(TRY300)